### ISSUE TITLE
Investigate use of spaCy for parsing

### ISSUE THREAD

**Issue by jzohrab** [jzohrab/lute-v3#43]
**This is a complicated issue**

spaCy and the stanford stanza project are very good parsing libraries, it would be nice to use something like that instead of (hacky?) regex solutions.  Unfortunately spaCy is very slow, so things would need to change quite a lot to make it usable within Lute.

Currently Lute parses **very frequently**:

- when a new book is created
- when a page is open for reading
- when a new term is created

To get around the frequent parsing (for reading), Lute could:

- parse the book once, and store the tokens **and** token boundaries (zero-width strings) in the texts.TxText field.  Then, when reading, everything is already parsed
- if creating terms from the reading pane, the zero-width strings (spans with spaces) could be sent to the route that creates the terms.  No extra parsing would be needed
- creating terms from the term index page would still require parsing ... that would be slow.  The parsing library could be loaded one-time only and then kept available for any runtime session
- I'm not sure how the spaCy dictionaries would be loaded, especially for Docker.  They'd have to be ... mounted somehow, in the user data folder.

**Comment by fanyingfx** [comment]
I have an idea, how about cache the parsed result by paragraph or sentence and do some persistence. Because lute can edit the text when reading , I not sure if it's suitable.

**Comment by jzohrab** [comment]
> cache the parsed result

Yep that's what is already said in the description.  There are other non-trivial challenges though, see the description.

**Comment by fanyingfx** [comment]
I see, In my understanding, it seems to the texts.TxText field cached result and can get from the route, I just misunderstood because I am not sure how to get the parsed result.

> creating terms from the term index page would still require parsing ... that would be slow. The parsing library could be loaded one-time only and then kept available for any runtime session

For this problem seem like make the parser instance as a singleton,  as I see in the the registry the same language parser instantiate multiple times, maybe can try this solution?  https://stackoverflow.com/a/73988733

**Comment by jzohrab** [comment]
> make the parser instance as a singleton

yes it would be something like that, that's what I meant with "the parsing library could be loaded one-time only and then kept available for any runtime session".  The current "registry" thing isn't the best implementation :-) but it does the job.

For spaCy, though, I'm not sure how it would perform, even if it were a singleton.  Having it loaded into memory could be very heavy, and it still might be slow.  Would need to test it out.  That's part of this ticket. :-)

> I just misunderstood because I am not sure how to get the parsed result.

That's in `lute/read/routes.py`, `def render_page(bookid, pagenum)`, where it calls `get_paragraphs(text)`, defined in `lute.read.service`.  It takes the text and parses it at runtime.  If the spaCy parser and its dictionaries etc can be stored in memory and can be run rapidly, then it might be sufficient.

Besides keeping the parser instance in memory, there is the question of how to get dictionaries/models.  SpaCy parsers require some kind of model, and if I recall correctly they're either installed with pip, or are somehow downloaded.  That creates other challenges, because Lute is pre-packaged, and with Docker it's even weirder.  I could be wrong, of course -- suffice it to say it's not a trivial problem.

**Comment by fanyingfx** [comment]
Yeah, I try to using the SpaCy to parser the Japanese, it has different ways to get model, some can installed by pip and some just using in the code then the SpaCy download it if needed, considering the docker it seems more complicated. Maybe there is a way to set the model path for saving and loading, but not sure, I can check it.

**Comment by fanyingfx** [comment]
I have tested this [solution ](https://stackoverflow.com/a/67750919) which can set the spacy model path for saving and can be loaded afterwards.  Hope it helps :-).
Below is the code: 
```python
cache_dir=os.getenv("SPACY_MODELS")
model_path="en_core_web_sm"
try:
    nlp = spacy.load(os.path.join(cache_dir,model_path))
except OSError:
    spacy.cli.download(model_path)
    nlp = spacy.load(model_path)
    nlp.to_disk(os.path.join(cache_dir,model_path))
```

**Comment by jzohrab** [comment]
> Hope it helps

Thank you :-)  I can't say when I'm going to look at this idea though, too much to do.  Cheers!
